### ISSUE TITLE
race-condition for ssh tunnel during `create-env`

### ISSUE THREAD

**Issue by voelzmo** [cloudfoundry/bosh-cli#110]
We've been seeing this quite frequently on certain installations, happening during `create-env` or with `bosh-init deploy`
```
Command 'deploy' failed:
  Deploying:
    Creating instance 'bosh/0':
      Updating instance disks:
        Updating disks:
          Deploying disk:
            Mounting disk:
              Sending 'get_task' to the agent:
                Agent responded with error: Action Failed get_task: Task 7e4d289d-b97c-4464-40d4-ecc90cc2a94b result: Persistent disk with volume id '14128b61-e046-48ae-b48a-fc0324716b83' could not be found
```

Which essentially means that the bosh-agent didn't find the disk UUID that gets passed in the `mount_disk` task in its settings. During `create-env` the agent reads the settings over a port forwarded in an ssh connection from the machine where you execute the bosh cli to the newly created VM. In the agent logs, we're seeing the registry-over-tunnel access work at some point in time:
```
2017-01-26_20:44:31.68555 [settingsService] 2017/01/26 20:44:31 DEBUG - Loading settings from fetcher
2017-01-26_20:44:32.19157 [registryProvider] 2017/01/26 20:44:32 DEBUG - Using http registry at http://127.0.0.1:6901
2017-01-26_20:44:33.22708 [settingsService] 2017/01/26 20:44:33 DEBUG - Successfully received settings from fetcher
```

but for the `mount_disk` task, it fails:
```
2017-01-26_20:49:02.45234 [Action Dispatcher] 2017/01/26 20:49:02 INFO - Received request with action mount_disk
2017-01-26_20:49:02.45234 [Action Dispatcher] 2017/01/26 20:49:02 DEBUG - Payload
2017-01-26_20:49:02.45235 ********************
2017-01-26_20:49:02.45235 {"method":"mount_disk","arguments":["45d691db-a656-4624-91a8-03ac746e8515"],"reply_to":"e169f631-7811-4faa-7693-98ccafabd8a4"}
2017-01-26_20:49:02.45236 [Action Dispatcher] 2017/01/26 20:49:02 INFO - Running async action mount_disk
2017-01-26_20:49:02.45236 [MBus Handler] 2017/01/26 20:49:02 INFO - Responding
2017-01-26_20:49:02.45237 [MBus Handler] 2017/01/26 20:49:02 DEBUG - Payload
2017-01-26_20:49:02.45237 ********************
2017-01-26_20:49:02.45237 {"value":{"agent_task_id":"7d9171e8-50cd-4df4-6a11-affbaf8843e3","state":"running"}}
2017-01-26_20:49:02.45238 ********************
2017-01-26_20:49:02.45265 [settingsService] 2017/01/26 20:49:02 DEBUG - Loading settings from fetcher
2017-01-26_20:49:04.03976 [settingsService] 2017/01/26 20:49:04 ERROR - Failed loading settings via fetcher: Getting settings from url: Get http://127.0.0.1:6901/instances/vm-79a0678d-87a5-4165-be61-ecab19807b20/settings: dial tcp 127.0.0.1:6901: getsockopt: connection refused
2017-01-26_20:49:04.04069 [File System] 2017/01/26 20:49:04 DEBUG - Reading file /var/vcap/bosh/settings.json
```

Which in turn [causes the whole task to fail with the above error message](https://github.com/cloudfoundry/bosh-agent/blob/master/agent/action/mount_disk.go#L56-L61), as the fallback settings from the filesystem don't contain the disk UUID (because the CPI just wrote them).

We've done some investigations and think it might be a race between the agent accessing the registry and the CLI closing down the tunnel:
* [`instance.WaitUntilReady()` creates the tunnel and defers a function that stops the tunnel](https://github.com/cloudfoundry/bosh-cli/blob/47ea06453290186de2610150498d3901d16c4131/deployment/instance/instance.go#L102-L106), which [closes the remote connection](https://github.com/cloudfoundry/bosh-cli/blob/master/deployment/sshtunnel/ssh_tunnel.go#L105)
* [`instance.WaitUntilReady()` is running until the agent could successfully be pinged](https://github.com/cloudfoundry/bosh-cli/blob/master/deployment/instance/instance.go#L114)
* Only after that, [`instance.manager` calls `UpdateDisks()`](https://github.com/cloudfoundry/bosh-cli/blob/47ea06453290186de2610150498d3901d16c4131/deployment/instance/manager.go#L128), `instance` calls [`vm.UpdateDisks()`](https://github.com/cloudfoundry/bosh-cli/blob/master/deployment/instance/instance.go#L126), and that [delegates to the `diskDeployer.deploy`](https://github.com/cloudfoundry/bosh-cli/blob/master/deployment/vm/vm.go#L135) which does the actual work of calling the CPI to [create](https://github.com/cloudfoundry/bosh-cli/blob/master/deployment/vm/disk_deployer.go#L100) and asking the `vm` to [attach](https://github.com/cloudfoundry/bosh-cli/blob/master/deployment/vm/disk_deployer.go#L105) 
* [`vm.attach` calls the CPI to attach the disk sends a `mount_disk` task to the agent](https://github.com/cloudfoundry/bosh-cli/blob/master/deployment/vm/vm.go#L148-L165)

By then, it seems, the connection is sometimes already closed. Wdyt?
I did an experiment with a modified version of `bosh-init` in which I removed the [deferred function](https://github.com/cloudfoundry/bosh-cli/blob/47ea06453290186de2610150498d3901d16c4131/deployment/instance/instance.go#L102-L106) and it worked nicely.

BTW while looking through the code, we thought that in case of an error during `accept()`, the tunnel will probably just hang, because of a blocking send on a channel `errCh` which nobody listens on: https://github.com/cloudfoundry/bosh-cli/blob/master/deployment/sshtunnel/ssh_tunnel.go#L49-L53

**Comment by cppforlife** [comment]
@voelzmo that's quite interesting. doesnt make sense to me why we would want to close tunnel at the end of WaitUntilReady(). that defer needs to be removed. somewhat surprised that we dont see this problem on AWS frequently (should be every deploy pretty much).

**Comment by voelzmo** [comment]
Interestingly enough, we only see this when craeting/attaching disks takes a longer time than usual. I'm also confused why we don't see this more often.
Maybe because the agent isn't closing every connection with an EOF immediately? Closing the remote connection just means that no new connections can be `Accept`ed, afaik it doesn't close already existing ones...

**Comment by voelzmo** [comment]
As agreed with @cppforlife: we're going to remove the  [deferred function](https://github.com/cloudfoundry/bosh-cli/blob/47ea06453290186de2610150498d3901d16c4131/deployment/instance/instance.go#L102-L106) that closes the remote connection in bosh-init and the bosh-cli.
Leave the error channel that nobody listens on alone for now.
